io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -723,30 +824,65 @@ reschedule !simstate@SimState{ threads, timers, curTime = time } =
 
         -- Reuse the STM functionality here to write all the timer TVars.
         -- Simplify to a special case that only reads and writes TVars.
-        !written <- execAtomically' (runSTM $ mapM_ timeoutAction fired)
-        (wakeup, wokeby) <- threadsUnblockedByWrites written
+        !written <- execAtomically' (runSTM $ mapM_ timeoutSTMAction fired)
+        (wakeupSTM, wokeby) <- threadsUnblockedByWrites written
         !_ <- mapM_ (\(SomeTVar tvar) -> unblockAllThreadsFromTVar tvar) written
 
-        let (unblocked,
-             simstate') = unblockThreads wakeup simstate
-        !trace <- reschedule simstate' { curTime = time'
-                                       , timers  = timers' }
+            -- Check all fired threadDelays
+        let wakeupThreadDelay = [ tid | TimerThreadDelay tid <- fired ]
+            wakeup            = wakeupThreadDelay ++ wakeupSTM
+            (_, !simstate')   = unblockThreads wakeup simstate
+
+            -- For each 'timeout' action where the timeout has fired, start a
+            -- new thread to execute throwTo to interrupt the action.
+            !timeoutExpired = [ (tid, tmid, isLockedRef)
+                              | TimerTimeout tid tmid isLockedRef <- fired ]
+
+        -- Get the isLockedRef values
+        !timeoutExpired' <- traverse (\(tid, tmid, isLockedRef) -> do
+                                        locked <- readSTRef isLockedRef
+                                        return (tid, tmid, isLockedRef, locked)
+                                     )
+                                     timeoutExpired
+
+        !simstate'' <- forkTimeoutInterruptThreads timeoutExpired' simstate'
+
+        !trace <- reschedule simstate'' { curTime = time'
+                                        , timers  = timers' }
+
         return $
-          traceMany ([ (time', ThreadId [-1], Just "timer", EventTimerExpired tmid)
-                     | tmid <- tmids ]
+          traceMany ([ ( time', ThreadId [-1], Just "timer"
+                       , EventTimerExpired tmid)
```

Comment:
Shouldn't this be `EventTimerFired`?

Reply:
I think I used "Expired" instead of "Fired" I can change that if you prefer